### PR TITLE
solve time issue when no RTC is present and (possibly) solve issue #119

### PR DESCRIPTION
**Pull request by joverbee**
detect if no real time clock is present and revert to browser clock and (possibly) solved issue #119 (although I can't test it as I have no RTC installed)

### CODE REVIEW

**Comment by joverbee** [comment]
Ok, working on it.
Is there a possibility to debug the js code without having to upload it to the board? IF I make a mistake now I have to open the evse box to hardwire to the wifi board. It seems there could be a smarter way?
But i guess this would need a part of the code acting as a fake evse board.

**Comment by joverbee** [comment]
screenshot 2018-01-21 17 09 32

Getting closer, can someone test if all still works with RTC and if issue #119 is still present?

**Comment by jeremypoulter** [comment]
I normally use either the openevse_ota config to flash the OpenEVSE over WiFi or you can run the files from the local disk by changing the baseHost at the top of home.js

**Comment by joverbee** [comment]
seems like it still has issues, sorry for the confusion

**Comment by joverbee** [comment]
this should be better...
Debugging this was particularly difficult as I don't get any compile errors or debug capabilities, only the output on the webpage is shown and when it locks up I needed to upload via usb. But when usb is connected, the evse is not running and therefore I can't check if the output is correct.

I tried changing the basehost=openevse.local in home.js, but then the local webpage opens but keeps waiting to load. Apparently I am still doing something wrong.
Does the OTA option have the same issue as the upload of firmware through the webpage? I.e. when the program hangs it also doesn't provide an upload function anymore, and this is exactly when you would need one ;).

I hope my approach is not too ridiculous.

**Comment by glynhudson** [comment]
Hi, I've just tested with a unit that has got a RTC present and it reports "no RTC detected"

selection_173

The RAPI command `$GT` reports the current time from the RTC:

selection_172

Is there anything I can do to help you debug?

**Comment by joverbee** [comment]
hmm, not good.

I've been breaking my head over it an trying different things:
raise an error when detecting that no RTC is present in openevse.js seems not unlogical
`
if (year==165 && month==165 && day==165 && hour==165 && minute==165 && second==85){
            request._error(new OpenEVSEError("No RTC"));
}
`
but now we need to catch this error in openevseviewmodel.js, for instance I thought this should work

``
self.hasRTC = ko.observable(true);
  self.result=self.openevse.time(self.RTCcallback);
  if (self.result.error.type=="No RTC"){
    self.hasRTC = ko.observable(false);
  }
  self.RTCcallback = function (date) {
  };``

but it doesnt and I am running out of ideas, I notice the RTCcallback never gets called, and whatever I tried the error seems not raised or I can't access it.

The good news:
manually setting self.hasRTC to false does result in the right behaviour in home.htm.
and I found out that locally running home.htm works after I realised I needed to take the compiled version in the 'data' folder after chaning the host in home.js...thanks @jeremypoulter for pointing me in this direction.

any ideas? (yes, I admit, I never programmed in js before)

**Comment by jeremypoulter** [comment]
See my previous comment, I would call the callback with false for no RTC.
Then you can use false === xxx to detect if there is no RTC.

Jeremy

On 23 Jan 2018 21:36, "Jo Verbeeck" <notifications@github.com> wrote:

hmm, not good.

I've been breaking my head over it an trying different things:
raise an error when detecting that no RTC is present in openevse.js seems
not unlogical
if (year==165 && month==165 && day==165 && hour==165 && minute==165 &&
second==85){ request._error(new OpenEVSEError("No RTC")); }
but now we need to catch this error in openevseviewmodel.js, for instance I
thought this should work

self.hasRTC = ko.observable(true);
self.result=self.openevse.time(self.RTCcallback);
if (self.result.error.type=="No RTC"){ self.hasRTC = ko.observable(false);
} self.RTCcallback = function (date) { };

but it doesnt and I am running out of ideas, I notice the RTCcallback never
gets called, and whatever I tried the error seems not raised or I can't
access it.

The good news:
manually setting self.hasRTC to false does result in the right behaviour in
home.htm.
and I found out that locally running home.htm works after I realised I
needed to take the compiled version in the 'data' folder after chaning the
host in home.js...thanks @jeremypoulter <https://github.com/jeremypoulter>
for pointing me in this direction.

any ideas? (yes, I admit, I never programmed in js before)

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/OpenEVSE/ESP8266_WiFi_v2.x/pull/132#issuecomment-359938786>,
or mute the thread

**Comment by joverbee** [comment]
Thanks for the suggestion, however I am confused:
I didn't try this at first  as I couldn't understand how a bool (false) could be converted into a date object.

Now I am trying either this
`callback(false)`
or this
```
date=false
callback(date)
```
But both seem to fail (the webpage doesnt load)
I'll keep searching, but hints are welcome.

**Comment by jeremypoulter** [comment]
Through the beauty/evil (depending on your perspective) of JavaScript.
Types are not fixed so you can return anything.

Both of your examples are correct, you will just have to change the higher
levels to handle the different data type use false === xxxx to test for
false being returned.

Jeremy

On 24 Jan 2018 19:21, "Jo Verbeeck" <notifications@github.com> wrote:

Thanks for the suggestion, however I am confused:
I didn't try this at first as I couldn't understand how a bool (false)
could be converted into a date object.

Now I am trying either this
callback(false)
or this

date=false
callback(date)

But both seem to fail (the webpage doesnt load)
I'll keep searching, but hints are welcome.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/OpenEVSE/ESP8266_WiFi_v2.x/pull/132#issuecomment-360244345>,
or mute the thread

**Comment by joverbee** [comment]
ok, I think it should work now
I changed the callback function to allow for a second bool to be passed (optional argument) 
`callback(date,false)`
it seems to work fine now as if I deliberately change to callback(date,true) I get back the 1970 date that agrees with date(0) on the webpage while if I do callback(date,false) it shows 'no RTC detected'.
So I assume it will correctly behave on an RTC system

**Comment by joverbee** [comment]
I would be more on the side of thinking that it is horrible if types are not fixed, because it could lead to all kinds of mistakes not being detected by the compiler and resulting in very lengthy debugging before realising where the problem is. But undoubtedly it also allows for more flexibility ;)
In any case, I am learning...

File: src/html/openevse.js [hunk 1]
```diff
@@ -185,8 +184,14 @@ function OpenEVSE(endpoint)
...
         if(!isNaN(year) && !isNaN(month) && !isNaN(day) && !isNaN(hour) && !isNaN(minute) && !isNaN(second)) {
-          var date = new Date(2000+year, month-1, day, hour, minute, second);
-          callback(date);
+          if (year==165 && month==165 && day==165 && hour==165 && minute==165 && second==85){
+            var date = new Date(0);
+            callback(date,false);//this pattern occurs when no RTC is connected to openevse
+          }
+          else{
+            var date = new Date(2000+year, month, day, hour, minute, second);
+            callback(date);
```
Comment (jeremypoulter) [inline src/html/openevse.js]: This case should return true, should not depend on the callback setting a default.

**Comment by jeremypoulter** [comment]
Looks good, gave it a quick test, my device with RTC is working as expected. Just one thing, can you target this PR at the master branch rather than stable? (On the PR, click 'Edit', then select master from the dropdown)

**Comment by joverbee** [comment]
ok done.
Does it actually solve issue #119? I can't test it.

**Comment by glynhudson** [comment]
Yes! Preliminary indicates this fixes #119. Nice work 👍 

I'll do some more testing and take a proper look next week before merging.

**Pull request merged**